src/analyzer/protocol/ssl/SSL.cc

```diff
@@ -98,4 +174,166 @@ void SSL_Analyzer::Undelivered(uint64_t seq, int len, bool orig)
 	interp->NewGap(orig, len);
 	}
 
+void SSL_Analyzer::SetSecret(const u_char* data, int len)
+	{
+	// FIXME: Is this the proper way to initialize a zeek::StringVal?
+	//        the alternative requires a cast: new zeek::StringVal(len, data)
+	secret = new zeek::StringVal(new zeek::String(data, len, true));
+	}
+
+void SSL_Analyzer::SetKeys(const u_char* data, int len)
+	{
+	keys = new zeek::StringVal(new zeek::String(data, len, true));
+	}
+
+bool SSL_Analyzer::TryDecryptApplicationData(int len, const u_char* data, bool is_orig, uint8_t content_type, uint16_t raw_tls_version)
+	{
+	// Unsupported cipher suite. Currently supported:
+	// - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 == 0xC030
+	auto cipher = handshake_interp->chosen_cipher();
+	if ( cipher != 0xC030 )
+		{
+		//printf("Unsupported cipher suite: %d\n", cipher);
+		return false;
+		}
+
+	// Neither secret or key present: abort
+	if ( secret->Len() == 0 && keys->Len() == 0 )
+		{
+		// FIXME: this is just for debugging
+		printf("Could not decrypt packet (missing key):\n");
+		print_hex("->client_random:", handshake_interp->client_random().data(), handshake_interp->client_random().length());
+		return false;
+		}
+
+	// Secret present, but no keys derived yet: derive keys
+	if ( secret->Len() != 0 && keys->Len() == 0 )
+		{
+		uint32_t ts = htonl((uint32_t) handshake_interp->gmt_unix_time());
+
+		u_char crand[32] = {0x00};
+		u_char keybuf[72];
+
+		auto c_rnd = handshake_interp->client_random();
+		auto s_rnd = handshake_interp->server_random();
+		memcpy(crand, &(ts), 4);
+		memcpy(crand + 4, c_rnd.data(), c_rnd.length());
+
+		tls1_prf(secret->Bytes(), secret->Len(), "key expansion", s_rnd.data(), s_rnd.length(),
+				crand, sizeof(crand), keybuf, sizeof(keybuf));
+
+		// save derived keys
+		SetKeys(keybuf, sizeof(keybuf));
+	}
+
+	// Keys present: decrypt TLS application data
+	if ( keys->Len() != 0 )
```

Comment:
this probably also should check that keys has the right length? I know that that is given at the moment - but I am a fan of defensive programming :)